### PR TITLE
Add DeriveKeyPair and separate serialization code from main protocol routines

### LINKED ISSUE #306: Specify key derivation
The current spec punts on key derivation. In particular, it assumes some mechanism for deriving keys. To promote an EdDSA-like interface, wherein "private keys" are random seeds used to _derive_ actual scalars, it would be useful to specify the key generation mechanism. This might look something like the following:

~~~
def DeriveKeyPair(seed, mode, suiteID, info):
  contextString = "VOPRF08-" || 
                  I2OSP(mode, 1) || I2OSP(suiteID, 2) ||
                  I2OSP(len(info), 2) || info
  skS = GG.HashToScalar(seed, DST = "DeriveKeyPair" || contextString)
  if skS == 0:
    raise DeriveKeyPairError
  pkS = ScalarBaseMult(skS)
  return skS, pkS
~~~

One could use this to derive keys for the OPRF, VOPRF, or POPRF protocol variants. For the POPRF case, `info` is repeated here and in the Evaluate interface, but this is not too dissimilar from HPKE, which admits application-specific information in the key schedule and at the AEAD level. Applications can choose how to expose `info` accordingly.

### PR DESCRIPTION
**Pull request by chris-wood**
Closes #306.

[Rendered version of this PR](https://cfrg.github.io/draft-irtf-cfrg-voprf/caw/dkp-with-serialization/draft-irtf-cfrg-voprf.html)

cc @bytemare, @kevinlewi, @tomrist

### CODE REVIEW

File: draft-irtf-cfrg-voprf.md [hunk 1]
```diff
@@ -762,6 +762,42 @@ def CreateContextString(mode, suiteID):
...
+  contextString = CreateContextString(mode, suiteID)
+  deriveInput = seed || I2OSP(len(info), 2) || info
+  counter = 0
+  skS = 0
+  while skS == 0:
+    if counter > 255:
+      raise DeriveKeyPairError
+    skS = GG.HashToScalar(deriveInput || I2OSP(counter, 1),
+                          DST = "DeriveKeyPair" || contextString)
+    counter = counter + 1
```
Comment (daxpedda) [inline draft-irtf-cfrg-voprf.md:796]: Is there interest in making this function infallible?

IANA cryptographer, but could `HashToScalar` be changed to always return a non-zero scalar by using the reduction mod n-1 plus one method. My assumption here is actually that `HashToScalar` returning zero scalars is actually a problem in other functions.

@armfazh mentions here https://github.com/cfrg/draft-irtf-cfrg-opaque/issues/298#issuecomment-1006936162 that `HashToScalar` shouldn't be subject to change, may I ask why?

I made a list of problems here https://github.com/cfrg/draft-irtf-cfrg-voprf/issues/307 that are affected by `HashToScalar` returning a zero scalar.

Comment (chris-wood) [reply draft-irtf-cfrg-voprf.md:788]: The probably that this fails is negligible, so it's pretty reasonable for implementations to panic if this occurs. In any case, we'll address fallible functions separately, not in this PR.

Comment (armfazh) [reply draft-irtf-cfrg-voprf.md:788]: > @armfazh mentions here [cfrg/draft-irtf-cfrg-opaque#298 (comment)](https://github.com/cfrg/draft-irtf-cfrg-opaque/issues/298#issuecomment-1006936162) that `HashToScalar` shouldn't be subject to change, may I ask why?

We would like the hash function behaves as a random oracle to the entire domain, in this case the whole field including zero.
This is similar to a regular hash function, for example, we cannot require that SHA-256 must not return 0.

**Comment by armfazh** [comment]
We can avoid the counter overflows by using an iterated hashing instead.
what about something like this?
```python
1. T = 000000
2. contextString = CreateContextString(mode, suiteID)
3. deriveInput = seed || I2OSP(len(info), 2) || info
4. MSG = deriveInput || T
5. DST = "DeriveKeyPair" || contextString
6. skS = GG.HashToScalar(MSG, DST)  
7. if skS == 0
8.    T = Hash(T)
9.    return to Step 4.
A. pks = ScalarBaseMult(skS)
B. return (skS,pkS).
```

**Comment by chris-wood** [comment]
> We can avoid the counter overflows by using an iterated hashing instead.

Probability of counter overflow is negligible, so I don't think we need to be overly clever with the construction here. It seems best to match key generation routines used for other algorithms like [HPKE](https://cfrg.github.io/draft-irtf-cfrg-hpke/draft-irtf-cfrg-hpke.html#name-derivekeypair). I suggest we stick with what's here in this PR, and file this iterated hash variant as an issue to possibly do later on.

File: draft-irtf-cfrg-voprf.md (hunk 1 above)
Comment (daxpedda) [reply draft-irtf-cfrg-voprf.md:788]: > In any case, we'll address fallible functions separately, not in this PR.

Sounds good!

**Review by armfazh — APPROVED** [review]

**Pull request merged**